Support for General Reports in AMMEX #2138

velmuruganc opened the pull request:
General Reports to AMMEX #2138

velmuruganc commented:
@wolfsolver , please review this pull. As of now, I have implemented the report view in HTML.

Also by mistake, I have checked in below files and it is same as what I have it in the PULL-2205. Plz ignore this file in this checkin and the correct one is in PULL/2205

app/src/main/java/com/money/manager/ex/notifications/SmsReceiverTransactions.java

guanlisheng commented:
this is good. for the code itself, can you take a look at the static code analysis results?

velmuruganc commented:
@wolfsolver , I fixed all static code analysis results except SMS related. For now, you can ignore this error. I have changed my method return from boolean to string and testing now. This will get corrected in my next pull request which will happen after covering all my SMS related testing (~75 cases to test).

guanlisheng commented:
hi @velmuruganc, can you further look at the conflicts after #2205 merged.  

additionally,  we may consider to impl a bi-sync between the user database.report_v1 and https://github.com/moneymanagerex/general-reports as the SUID is in place shortly.

velmuruganc commented:
> hi @velmuruganc, can you further look at the conflicts after #2205 merged.
> 

@guanlisheng , I resolved the conflict.

guanlisheng reviewed (approved):
lgtm.  @wolfsolver, thought?

wolfsolver commented:
@velmuruganc seems that you reopen db from drawer instead reuse db connection
```
        MmxOpenHelper MmxHelper = new MmxOpenHelper(this, new AppSettings(this).getDatabaseSettings().getDatabasePath());
        SupportSQLiteDatabase db = MmxHelper.getReadableDatabase();
```

@guanlisheng  what is the best practice for this? register new URI?

wolfsolver commented on app/src/main/java/com/money/manager/ex/home/MainActivity.java:
```diff
@@ -1412,4 +1436,116 @@ private void showSelectDatabaseActivity() {
+        ArrayList<DrawerMenuItem> childReportGroup = new ArrayList<>();
+
+        // Db setup
+        MmxOpenHelper MmxHelper = new MmxOpenHelper(this, new AppSettings(this).getDatabaseSettings().getDatabasePath());
```
I think this is not the best way to get dB data.

Create a content provider and register with Uri, so you can use mmx infrastructure to get data

wolfsolver commented on app/src/main/java/com/money/manager/ex/home/MainActivity.java:
```diff
@@ -1412,4 +1436,116 @@ private void showSelectDatabaseActivity() {
+
+        // Db setup
+        MmxOpenHelper MmxHelper = new MmxOpenHelper(this, new AppSettings(this).getDatabaseSettings().getDatabasePath());
+        SupportSQLiteDatabase db = MmxHelper.getReadableDatabase();
```
Same as above

wolfsolver commented on app/src/main/java/com/money/manager/ex/home/MainActivity.java:
```diff
@@ -1412,4 +1436,116 @@ private void showSelectDatabaseActivity() {
+        int iconColor = uiHelper.getSecondaryTextColor();
+
+        // Db setup
+        MmxOpenHelper MmxHelper = new MmxOpenHelper(this, new AppSettings(this).getDatabaseSettings().getDatabasePath());
+        SupportSQLiteDatabase db = MmxHelper.getReadableDatabase();
```
Same as above . You reopen dB too many times

wolfsolver commented on app/src/main/java/com/money/manager/ex/reports/GeneralReportFragment.java:
```diff
@@ -0,0 +1,134 @@
+    private String getHtmlReport(String reportName){
+
+        //Db setup
+        MmxOpenHelper MmxHelper = new MmxOpenHelper(getActivity(), new AppSettings(getActivity()).getDatabaseSettings().getDatabasePath());
+        SupportSQLiteDatabase mDatabase = MmxHelper.getReadableDatabase();
```
Same as above... Don't reopen db

wolfsolver commented on app/src/main/java/com/money/manager/ex/reports/GeneralReportFragment.java:
```diff
@@ -0,0 +1,134 @@
+
+                if(sqlCursor.moveToFirst())
+                {
+                    sqlQuery = sqlCursor.getString(sqlCursor.getColumnIndex("SQLCONTENT"));
```
I think that general report have also Lua content for elaborate data. 
Did you plan do add also this feature?

velmuruganc replied:
Yes, I am going to explore next

wolfsolver commented on app/src/main/java/com/money/manager/ex/home/MainActivity.java:
```diff
@@ -1055,6 +1070,14 @@ private ArrayList<DrawerMenuItem> getDrawerMenuItems() {
         // .withDivider(true));
+
+        // General reports
+        menuItems.add(new DrawerMenuItem().withId(R.id.menu_general_report_group)
```
Better to enable this only if exist report on DB?
```
if (generalReport.exisit()) {
  menuitem.add(....
} 
```

wolfsolver reviewed (commented):
See my comment inline code

velmuruganc commented:
> Member

@wolfsolver , thankyou so much for the feedback. I will look into these comments and get back to you. 

Lot to learn from you both.

wolfsolver commented:
a short way, without changing a lot is to include this as global variable. in this way to get the already instiated variable. not the best way, but quick for now. and remove new isntance declaration
```
    @Inject
    Lazy<MmxOpenHelper> openHelper;
```

velmuruganc commented:
> one more tiny thing, would we keep original `Reports` unchanged?

You mean report name?
